File: source/ts/src/core/rashi.ts
import { nakshatraOf } from '../utils/constants';
import type { RashiInfo, NakshatraIndexInfo } from '../types/elements';

export function computeChandraRashi(
  siderealMoon: number,
  nameFn: (index: number) => string,
): RashiInfo {
  const index = Math.floor(siderealMoon / 30);
  return { index, name: nameFn(index) };
}

export function computeSuryaNakshatra(
  siderealSun: number,
  nameFn: (index: number) => string,
): NakshatraIndexInfo {
  const index = nakshatraOf(siderealSun);
  return { index, name: nameFn(index) };
}

File: source/go/internal/core/rashi.go
package core

import (
	"math"

	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

func ComputeChandraRashi(siderealMoon float64, nameFn func(index int) string) types.RashiInfo {
	index := int(math.Floor(siderealMoon / 30))
	return types.RashiInfo{Index: index, Name: nameFn(index)}
}

func ComputeSuryaNakshatra(siderealSun float64, nameFn func(index int) string) types.NakshatraIndexInfo {
	index := utils.NakshatraOf(siderealSun)
	return types.NakshatraIndexInfo{Index: index, Name: nameFn(index)}
}
